Primary shard allocator observes limits in forcing allocation

abeyad opened the pull request:
Previously, during primary shards allocation of shards
with prior allocation IDs, if all nodes returned a
NO decision for allocation (e.g. the settings blocked
allocation on that node), we would chose one of those
nodes and force the primary shard to be allocated to it.

However, this meant that primary shard allocation
would not adhere to the decision of the MaxRetryAllocationDecider,
which would lead to attempting to allocate a shard
which has failed N number of times already (presumably
due to some configuration issue).

This commit solves this issue by introducing the
notion of force allocating a primary shard to a node
and each decider implementation must implement whether
this is allowed or not. In the case of MaxRetryAllocationDecider,
it just forwards the request to canAllocate.

Closes #19446

abeyad commented:
@ywelsch FYI, initial pass on this, would welcome your feedback.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java:
```diff
@@ -98,4 +98,16 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
+    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard routing [" +
+                                            shardRouting.shardId() + "]";
+        return Decision.YES;
```
this means that throttling is just ignored.
Assume scenario where FilterAllocationDecider says NO and ThrottlingAllocationDecider says THROTTLE. The implementation here would just forcefully allocate the shard, ignoring the throttling.

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -268,7 +290,9 @@ private boolean isEnoughVersionBasedAllocationsFound(IndexMetaData indexMetaData
-    private NodesToAllocate buildNodesToAllocate(ShardRouting shard, RoutingAllocation allocation, List<NodeGatewayStartedShards> nodeShardStates) {
+    private NodesToAllocate buildNodesToAllocate(RoutingAllocation allocation,
+                                                 List<NodeGatewayStartedShards> nodeShardStates,
+                                                 Function<RoutingNode, Decision> decisionFunction) {
```
we could alternatively just pass a boolean `forceAllocate` instead of the decisionFunction.

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -154,17 +154,39 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
-                changed = true;
-                unassignedIterator.initialize(nodeShardState.getNode().getId(), nodeShardState.allocationId(), ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE);
+                // We have zero nodes for which a decision of yes or throttle was given, so
+                // instead of iterating forcing allocation of the shard to one of the "no" nodes,
```
once this PR is in, there is no point in commenting about what should NOT be done ;-) I would just write: "All nodes with shard copies returned NO decision, check if primary shard can be forced to one of the nodes".

abeyad commented:
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/19811/commits/397e374e6f3cb01353fbea8b8380806085935aa9 to address your feedback

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java:
```diff
@@ -98,4 +98,21 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
+    /**
+     * Returns a {@link Decision} whether the given primary shard can be
+     * forcibly allocated on the given node. This method should only be called
+     * on nodes for which previous allocations exist for the primary shard.
```
"This method should only be called on nodes for which previous allocations exist for the primary shard." should be more like "This method should only be called for unassigned primary shards where the node has a shard copy on disk."
Can you also assert `shardRouting.unassigned()` ?

abeyad replied:
Done

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java:
```diff
@@ -196,4 +196,29 @@ public Decision canRebalance(RoutingAllocation allocation) {
+    @Override
+    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard routing [" +
+                                            shardRouting.shardId() + "]";
```
put shardRouting there, not only shardId (i.e. remove .shardId()).

ywelsch replied:
similar as for `canAllocate`, I think we should keep the condition

```
if (allocation.shouldIgnoreShardForNode(shardRouting.shardId(), node.nodeId())) {
            return Decision.NO;
}
```

abeyad replied:
Done, good catch

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -154,17 +154,34 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
+                );
+                if (nodesToForceAllocate.yesNodeShards.isEmpty() == false) {
+                    NodeGatewayStartedShards nodeShardState = nodesToForceAllocate.yesNodeShards.get(0);
+                    logger.debug("[{}][{}]: force allocating [{}] to [{}] on primary allocation",
```
I would change the message to read `[{}][{}]: allocating [{}] to [{}] on forced primary allocation`

abeyad replied:
Done

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -154,17 +154,34 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
+                    unassignedIterator.initialize(nodeShardState.getNode().getId(), nodeShardState.allocationId(),
+                                                  ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE);
+                } else if (nodesToForceAllocate.throttleNodeShards.isEmpty() == false) {
+                    logger.debug("[{}][{}]: force throttling allocation [{}] to [{}] on primary allocation",
```
same as above, I would change message to `[{}][{}]: throttling allocation [{}] to [{}] on forced primary allocation`

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -154,17 +154,34 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
+                                 shard.index(), shard.id(), shard, nodesToForceAllocate.throttleNodeShards);
+                    changed |= unassignedIterator.removeAndIgnore(AllocationStatus.DECIDERS_THROTTLED);
+                } else {
+                    logger.debug("[{}][{}]: primary allocation denied [{}]", shard.index(), shard.id(), shard);
```
`[{}][{}]: forced primary allocation denied [{}]`

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/MaxRetryAllocationDeciderTests.java:
```diff
@@ -208,4 +213,45 @@ public void testFailedAllocation() {
+     * Tests that {@link MaxRetryAllocationDecider#canForceAllocatePrimary(ShardRouting, RoutingNode, RoutingAllocation)} adheres
+     * to the maximum retry count and returns a NO decision if the maximum number of retries is exceeded.
+     */
+    public void testCanForceAllocatePrimary() {
```
If you just want to check that `canForceAllocatePrimary` adheres to the maximum retry count, isn't it simpler to extend the existing test `testFailedAllocation` and 
do a simple check there? Such a check can be made by selecting the unassigned shard in the routing table and doing 
```
assertEquals(Decision.NO, new MaxRetryAllocationDecider(Settings.EMPTY).canForceAllocatePrimary(shardRouting, null, new RoutingAllocation(null, null, clusterState, null, 0, false)))
```

abeyad replied:
I removed this test and put the canForceAllocatePrimary checks in `testFailedAllocation`

ywelsch commented on core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java:
```diff
@@ -193,6 +199,52 @@ public void testFoundAllocationAndAllocating() {
+        assertEquals(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).size(), 1);
+        assertEquals(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(), node1.getId());
+
+        // tests that we do not force allocate the primary when at least one decider returns NO or THROTTLE for canForceAllocatePrimary
```
can we have different test methods here for the different parts that are test here?

abeyad replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java:
```diff
@@ -193,6 +199,52 @@ public void testFoundAllocationAndAllocating() {
+
+        // tests that we do not force allocate the primary when at least one decider's decision is to THROTTLE
+        deciders = new AllocationDeciders(Settings.EMPTY, new AllocationDecider[] {
+            new TestAllocateDecision(Decision.THROTTLE), new ForcePrimaryDecider(Decision.YES)
```
is this really testing the right thing? This test does not seem to call `canForceAllocatePrimary` as `TestAllocateDecision` returns THROTTLE and not NO for `canAllocate`.

abeyad replied:
Yes, because the `ForcePrimaryDecider` here returns `NO` for its canAllocate decision.  As we discussed, its still a bit confusing, so I explicitly added the canAllocate decision as another constructor argument to `ForcePrimaryDecider` and added comments to explain whats happening.

ywelsch commented:
@abeyad I've left some comments. Can you also check if there are tests that `FilterAllocationDecider` and other deciders are indeed allowing force-allocating primaries?

I would also like to see some docs where we can explain the implemented behavior. This could be useful to understand why primary shards are / are not allocated.

abeyad commented:
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/19811/commits/9e47cd14abc3ff03bac4c32616a6b2b2e502f8ce, which addresses your review comments, and adds `PrimaryAllocationIT` test that covers filter allocation decider force allocating primaries.

As we discussed, I augmented the `PrimaryShardAllocator` with some more javadocs to explain the behavior.  Docs explaining shard allocation in general (including primary shard allocation) will come in a separate PR.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java:
```diff
@@ -98,4 +98,15 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
+    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard " + shardRouting;
+        assert shardRouting.unassigned() : "must not call canForceAllocatePrimary on an assigned shard " + shardRouting;
+        return Decision.YES; // by default, a decider will allow force allocation of the primary
```
I think I prefer the way it was before with the default being 

```
decision = canAllocate(...)
if (decision == Decision.NO) {
    decision = Decision.single(Type.YES, "force override of " + decision.label, ...)
}
```

This keeps the override logic out of `AllocationDeciders`.

abeyad replied:
My concern with your suggestion (which was the initial approach) is that for any decider that overrides `canForceAllocatePrimary`, it is easy for it to forget to take into account the decision of `canAllocate`.  That's why I figured it would be better to put the invocation of `canAllocate` in `AllocationDeciders`, then if the decision is `NO` (as opposed to throttle, for example), then call the deciders `canForceAllocatePrimary` implementation.  The only downside to this that I can think of is that its rigid in that no `canForceAllocatePrimary` method would be able to override a non-NO `canAllocate` decision - but I don't see any reason why we would want that?

abeyad replied:
I pushed 5c25b15c7a33ffa6b9e0965f30694f3f13bcd82a to address this

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/ESAllocationTestCase.java:
```diff
@@ -196,6 +196,32 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
+     * A decider class that allows controlling the decision on both allocating (AllocationDecider#canAllocate)
+     * and force allocating (AllocationDecider#canForceAllocatePrimary) a primary shard.
+     */
+    protected static class ForcePrimaryDecider extends TestAllocateDecision {
```
Instead of having this class, it might be more readable to just have two methods in `PrimaryShardAllocatorTests`:

```
AllocationDecider getNoDeciderThatThrottlesForceAllocate() {
   return new TestAllocateDecision(Decision.NO) {
       @Override
       public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
            return Decision.THROTTLE;
       }
   };
}

AllocationDecider getNoDeciderThatAllowsForceAllocate() {
     ... return YES;

}
```

abeyad commented:
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/19811/commits/9446267af3b1b938ebbfb0ff39916c9d9186009e to remove `ForcePrimaryDecider` in favor of anonymous class creation that extends `TestAllocateDecision`

abeyad commented:
@elasticmachine retest this please

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/MaxRetryAllocationDecider.java:
```diff
@@ -80,4 +80,12 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocat
+
+    @Override
+    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard [" + shardRouting.shardId() + "]";
```
print `shardRouting`, not only `shardRouting.shardId()`.

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -154,17 +166,37 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
-                unassignedIterator.initialize(nodeShardState.getNode().getId(), nodeShardState.allocationId(), ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE);
+                // The deciders returned a NO decision for all nodes with shard copies, so we check if primary shard
+                // can be force-allocated to one of the nodes, despite getting a NO decision for allocation. The decision
+                // to force allocate a primary to a node comes from AllocationDeciders#canForceAllocatePrimary, which calls
```
this is not strictly true anymore with the last change? update docs here.

abeyad replied:
I think the comment is a bit confusing in general so I'll remove it since the behavior is clear from the javadocs for `canForceAllocatePrimary`

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java:
```diff
@@ -219,4 +220,33 @@ public void testNotWaitForQuorumCopies() throws Exception {
+        final String indexName = "test-idx";
+        assertAcked(client().admin().indices()
+                        .prepareCreate(indexName)
+                        .setWaitForActiveShards(ActiveShardCount.ALL)
```
we get this for free with 0 replicas, no need to add this.

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java:
```diff
@@ -219,4 +220,33 @@ public void testNotWaitForQuorumCopies() throws Exception {
+        internalCluster().fullRestart();
+        logger.info("--> checking that the primary shard is force allocated to the data node despite being blocked by the exclude filter");
+        ensureYellow(indexName);
+        assertEquals(1, internalCluster().clusterService(internalCluster().getMasterName())
```
why explictly use master node? just make a proper client request (`client.admin().cluster().prepareState().get()...`)

ywelsch commented on core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java:
```diff
@@ -193,6 +198,74 @@ public void testFoundAllocationAndAllocating() {
+        AllocationDeciders deciders = new AllocationDeciders(Settings.EMPTY, new AllocationDecider[] {
+            // since both deciders here return a NO decision for allocating a shard,
+            // the allocator will see if it can force assign the primary, where the decision will be YES
+            new TestAllocateDecision(Decision.NO), getNoDeciderThatAllowsForceAllocate()
```
does it make sense to replace `new TestAllocateDecision(Decision.NO)` by `new TestAllocateDecision(randomBoolean() ? Decision.YES : Decision.NO)`.

ywelsch commented on core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java:
```diff
@@ -193,6 +198,74 @@ public void testFoundAllocationAndAllocating() {
+    public void testDontForceAllocateOnThrottleDecision() {
+        testAllocator.addData(node1, ShardStateMetaData.NO_VERSION, "allocId1", randomBoolean());
+        AllocationDeciders deciders = new AllocationDeciders(Settings.EMPTY, new AllocationDecider[] {
+            // since both deciders here return a NO decision for allocating a shard,
```
this comment is not true here.

ywelsch commented on core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java:
```diff
@@ -573,6 +647,36 @@ private void assertClusterHealthStatus(RoutingAllocation allocation, ClusterHeal
+        };
+    }
+
+    private AllocationDecider getNoDeciderThatDeniesForceAllocate() {
```
we can have three methods here, but all of them could just share their code by calling a common method `getNoDeciderWithForceAllocate(Decision)`

ywelsch commented:
@abeyad Left minor comments about docs/tests, change looks good o.w.

abeyad commented:
@ywelsch I pushed 5b536be443abb2aaf5a389cbd3066c6811c685ca

ywelsch commented:
LGTM. Thanks @abeyad!

abeyad commented:
Thanks for the review @ywelsch !
